File: desafio3/rust/src/main.rs
use std::env;
use std::error::Error;
use std::io::BufReader;
use std::io::BufRead;
use std::io::BufWriter;
use std::io::Write;
use std::fs::File;
use std::str;
use std::collections::BTreeSet;

const POS_VECTOR: usize = 9;
const ELEMENTOS_VECTOR: usize = 23;
const TAM_PERIODO: usize = 6;
const CANT_INSTITUCIONES: usize = 6;
const TAM_VECTOR: usize = ELEMENTOS_VECTOR * TAM_PERIODO;
const TAM_VECTOR_ENTRADA: usize = TAM_VECTOR * CANT_INSTITUCIONES;
const LARGO_LINEA: usize = POS_VECTOR + TAM_VECTOR_ENTRADA;
const TAM_SALIDA: usize = POS_VECTOR + 1 + TAM_VECTOR;

const PERIODO_NULO: [u8; TAM_PERIODO] = [b'0'; TAM_PERIODO];

fn ordenar_vector(vector:&[u8],  result:&mut [u8]) {

	let periodos : BTreeSet<&[u8]> = vector.chunks(TAM_PERIODO).filter(|p| p != &PERIODO_NULO).collect();

	// retorna el resultado
	if periodos.is_empty() {
		result[0] = b'N';
	} else if periodos.len() > ELEMENTOS_VECTOR {
		result[0] = b'S';
	} else {
		result[0] = b'D';
		let mut p = 1;
		for per in periodos.iter().rev() {
			result[p..p+TAM_PERIODO].clone_from_slice(&per);
			p += TAM_PERIODO;
		}
	}
}


fn procesar_linea(buf: &[u8]) -> [u8; TAM_SALIDA] {
	let mut res : [u8; TAM_SALIDA] = [b' '; TAM_SALIDA];
	res[0..POS_VECTOR].clone_from_slice(&buf[0..POS_VECTOR]);
	ordenar_vector(&buf[POS_VECTOR..], &mut res[POS_VECTOR..]); 
	res
}


fn main() {
	let args: Vec<String> = env::args().collect();
	if args.len() != 3 {
		println!("Uso: ordena_vector archivo_entrada archivo_salida");
		return
	} 

    let t0 = std::time::Instant::now();

	let entrada = match File::open(&args[1]) {
		Err(e) => panic!("No pudo abrir archivo {}, causa: {}", args[1], e.description()),
		Ok(file) => BufReader::with_capacity(64*1024, file)
	};
	let mut salida = match File::create(&args[2]) {
		Err(e) => panic!("No pudo crear archivo {}, causa: {}", args[2], e.description()),
		Ok(file) => BufWriter::with_capacity(64*1024, file)
	};

	for (num_linea, linea) in entrada.lines().enumerate() {
		let buf = &linea.unwrap().into_bytes();
		
		if buf.len() != LARGO_LINEA {
			println!("Error en el largo de la linea {}", num_linea);
			writeln!(salida, "{}", str::from_utf8(&buf).unwrap()).unwrap();
		} else {
			let vector = procesar_linea(buf);
			salida.write(&vector).unwrap();
			salida.write(b"\n").unwrap();
		};
	}
	let dur  = t0.elapsed();
	let secs = dur.as_secs();
    let frac = dur.subsec_millis();
    println!("tiempo ocupado: {}.{}s", secs, frac);
}

File: desafio3/scala/src/main/scala/Main.scala
import scala.io._
import scala.util._
import scala.collection.immutable.StringOps._
import scala.collection.mutable._
import java.io._

object Main {

  val bufferSize = 4096
  val encoding = "UTF-8"
  val time_report_message = "tiempo ocupado: "
  val tamPeriodo = 6
  val cantInstituciones = 6
  val tamVector = 23
  val posVector = 9
  val largoLinea = posVector + cantInstituciones * tamVector * tamPeriodo
  val tamRelleno = tamVector * tamPeriodo
  val ceros = "000000" // tamPeriodo 

  sealed trait Args
  case object BadArgs extends Args
  case class FileArgs(entrada:String, salida:String) extends Args

  def checkArgs(args: Array[String]) : Args = 
    if (args.size != 2) 
      BadArgs
    else 
        FileArgs(args(0), args(1))

  def ordenarPeriodos(linea: String) : String = {
    val encabezado = linea.slice(0, posVector)
    val myOrdering = Ordering.fromLessThan[String](_ > _)
    var periodos = SortedSet.empty[String](myOrdering)
    var pos = posVector
    while (pos < largoLinea) {
      if (!linea.regionMatches(pos, ceros, 0, tamPeriodo)) 
         periodos += linea.slice(pos, pos+tamPeriodo)
      pos += tamPeriodo
    }

    
    val len = periodos.size
    encabezado + (
      if (len == 0) 
          "N" + (" " * tamRelleno)
      else if (len > tamVector)
          "S" + " " * tamRelleno
      else 
          "D" + periodos.mkString + " "*(tamRelleno-len*tamPeriodo)
    )
  }

  def filtrarLinea(linea: String, nl: Long) : String = 
    if (linea.length() == largoLinea) 
      ordenarPeriodos(linea)
    else {
      println(f"!!! Largo incorrecto en linea: $nl%d\n")
      linea
    } 

  def procesarArchivos(entrada: String, salida: String) : Unit = {
    Try(Source.fromFile(new File(entrada), encoding, bufferSize)) match {
      case Failure(e) =>  println(f"No pudo abrir archivo $entrada%s, causa: $e%s")
      case Success(rdr) => 
        Try(new BufferedWriter(new OutputStreamWriter(new FileOutputStream(salida)))) match {
          case Failure(e) => println(f"No pudo crear archivo $salida%s, causa: $e%s")
          case Success(wrt) => 
            rdr.getLines.zipWithIndex.foreach { 
              case (linea, nl) =>
                wrt.write( filtrarLinea(linea, nl) )
                wrt.newLine()
            }
            wrt.close()
        }
    }
  }

  def showTime(block : => Unit) : Unit = {
    val t0 = System.nanoTime()
    val result = block
    val t1 = System.nanoTime()
    val elapsed = t1 - t0
    val msecs = elapsed / 1000000
    val secs = msecs / 1000.0

    println(f"$time_report_message $secs%02.3f")
  }

  def main(args: Array[String]): Unit = 
    checkArgs(args) match {
      case BadArgs => println("Uso: ordenar_vector archivo_entrada archivo_salida")
      case FileArgs(entrada, salida) => showTime { procesarArchivos(entrada, salida) }
    } 

}